src/database/database.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/vshn/cloudscale-metrics-collector/pkg/queriesmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/tenantsmodel"
+	"os"
+	"strings"
+	"time"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	dbUrlEnvVariable   = "ACR_DB_URL"
+	sourceQueryStorage = "object-storage-storage"
+	provider           = "exoscale"
+	queryAndZone       = sourceQueryStorage + ":" + provider
+	defaultUnit        = "GBDay"
+)
+
+var (
+	productRow = db.Product{
+		Source: queryAndZone,
+		Target: sql.NullString{String: "1402", Valid: true},
+		Amount: 0.00066,
+		Unit:   "GBDay",
+		During: db.InfiniteRange(),
+	}
+	discountRow = db.Discount{
+		Source:   sourceQueryStorage,
+		Discount: 0,
+		During:   db.InfiniteRange(),
+	}
+	queryRow = db.Query{
+		Name:        queryAndZone,
+		Description: "Object Storage - Storage (exoscale.com)",
+		Query:       "",
+		Unit:        "GBDay",
+		During:      db.InfiniteRange(),
+	}
+)
+
+// AggregatedBucket contains total used storage in an organization namespace
+type AggregatedBucket struct {
+	Organization string
+	// Storage in bytes
+	StorageUsed int64
+}
+
+// Database holds raw url of the postgresql database with the opened connection
+type Database struct {
+	Url        string
+	connection *sqlx.DB
+}
+
+// Configure loads environment variable which holds the database URL
+func Configure() (database *Database, err error) {
+	databaseUrl := os.Getenv(dbUrlEnvVariable)
+	if databaseUrl == "" {
+		return nil, fmt.Errorf("cannot find environment variable %s", dbUrlEnvVariable)
+	}
+	return &Database{Url: databaseUrl}, nil
+}
+
+// OpenConnection opens a connection to the postgres database
+func (d *Database) OpenConnection() error {
+	connection, err := db.Openx(d.Url)
+	if err != nil {
+		return fmt.Errorf("cannot create a connection to the database: %w", err)
+	}
+	d.connection = connection
+	return nil
+}
+
+// CloseConnection closes the connection to the postgres database
+func (d *Database) CloseConnection() error {
+	err := d.connection.Close()
+	if err != nil {
+		return fmt.Errorf("cannot close database connection: %w", err)
+	}
+	return err
+}
+
+// EnsureBucketUsage saves the aggregated buckets usage by namespace to the postgresql database
+// To save the correct data to the database the function also matches a relevant product, discount (if any) and query.
+// The storage usage is referred to a day before the application ran (yesterday)
+func (d *Database) EnsureBucketUsage(ctx context.Context, namespace string, aggregatedBucket AggregatedBucket, billingDate time.Time) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Saving buckets usage for namespace", "namespace", namespace, "storage used", aggregatedBucket.StorageUsed)
+
+	// start new transaction for actual work
+	tx, err := d.connection.BeginTxx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot create database transaction for namespace %s: %w", namespace, err)
+	}
```

Comment:
> Maybe also put this in a pipeline?

I understand that you want a solution for the ridiculous go error handling code, but at the same time I'm a bit torn on this. I feel that the pipeline library obscures the function of the code a bit too much and actually makes the readability of the code significantly worse. So personally I would avoid using it.

But again, I understand why you want something like that, and I don't really have strong feelings, so feel free to use it anyway.

Reply:
I might change it into a pipeline but I will probably need to use functions that return functions because of missing passing arguments around. I will play a bit with this and see if it makes sense.

Reply:
Moved to a pipeline. Now we have more methods and somewhat less clutter :)